[Quant][PT2E]Enable x86 inductor quantizer

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102516
* #98826
* #101941
* __->__ #98730

**Summary**

- Enable `X86InductorQuantizer` basics. 
- Recipe to annotate conv2d is added.

**Test Plan**
```
python -u -m pytest -s -v test_x86inductor_quantizer.py -k TestQuantizePT2EX86Inductor
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @Xia-Weiwen

leslie-fang-intel committed:
Enable x86 inductor quantizer

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

jerryzh168 reviewed (commented):
please wait a bit, there are a few renaming/restructuring PRs that I need to land, should be ready in 1-2 days

leslie-fang-intel commented:
> please wait a bit, there are a few renaming/restructuring PRs that I need to land, should be ready in 1-2 days

@jerryzh168 Thanks for the info. I will use this PR for internal testing for now. Please let me know when your renaming/restructuring PRs landed. I will rebase this PR and ask your kindly review after that.

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

Xia-Weiwen commented on test/quantization/fx/test_quantize_pt2e.py:
```diff
@@ -597,3 +598,56 @@ def test_resnet18_with_quantizer_api(self):
             self.assertTrue(torch.max(after_quant_result - after_quant_result_fx) < 1e-1)
             self.assertTrue(compute_sqnr(after_quant_result, after_quant_result_fx) > 35)
+
+@skipIfNoQNNPACK
```
Is this necessary? I think we can go without qnnpack, right?

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[WIP][Quant][PT2E]Enable x86 inductor quantizer"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Enable x86 inductor quantizer"


**Summary**

- Enable `X86InductorQuantizer` basics. 
- Recipe to annotate conv2d is added.

**Test Plan**
```
python -u -m pytest -s -v test_quantize_pt2e.py -k TestX86InductorQuantizePT2E
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

jgong5 commented on torch/ao/quantization/_pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -0,0 +1,205 @@
+    "get_default_x86_inductor_quantization_config",
+]
+
+def supported_quantized_operators() -> Dict[str, List[OperatorPatternType]]:
```
add TODO comment to add more supported operators?

leslie-fang-intel replied:
Thanks for the suggestion and added.

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1750,3 +1754,54 @@ def test_qat_resnet18(self):
                 m, example_inputs, is_per_channel=True,
             )
+
+@skipIfNoDynamoSupport
```
Lets start splitting tests. I would suggest `test_x86inductor_quantizer.py`

leslie-fang-intel replied:
Thanks, split the testing into `test_x86inductor_quantizer.py`.

kimishpatel commented on torch/ao/quantization/_pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -0,0 +1,205 @@
+        return list(op_configs)
+
+    @classmethod
+    def get_supported_operator_for_quantization_config(
+        cls, quantization_config: Optional[QuantizationConfig]
+    ) -> List[OperatorPatternType]:
+        if quantization_config is None:
+            all_ops = []
+            for _, ops in cls.supported_config_and_operators:
+                all_ops.extend(ops)
+            return all_ops
+
+        for config, ops in cls.supported_config_and_operators:
+            if config == quantization_config:
+                return ops
+        return []
```
we havent quite figured if we gonna keep this interface. cc: @jerryzh168

leslie-fang-intel replied:
Oh, yean. Should I remove it for now or should we keep aligned with `qnnpack_quantizer`?

jerryzh168 replied:
please wait a bit, we're planning to discuss this soon

kimishpatel commented on torch/ao/quantization/_pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -0,0 +1,205 @@
+        return model
+
+    def _annotate_conv2d(
+        self, gm: torch.fx.GraphModule, quantization_config: QuantizationConfig
+    ) -> None:
+        conv_partitions = get_source_partitions(
+            gm.graph, [torch.nn.Conv2d, torch.nn.functional.conv2d]
+        )
+        conv_partitions = list(itertools.chain(*conv_partitions.values()))
+        for conv_partition in conv_partitions:
+            if len(conv_partition.output_nodes) > 1:
+                raise ValueError("conv partition has more than one output node")
+            conv_node = conv_partition.output_nodes[0]
+            if (
+                conv_node.op != "call_function"
+                or conv_node.target != torch.ops.aten.convolution.default
+            ):
+                raise ValueError(f"{conv_node} is not an aten conv2d operator")
+            # skip annotation if it is already annotated
+            if _is_annotated([conv_node]):
+                continue
+            input_qspec_map = {}
+            input_node = conv_node.args[0]
+            assert isinstance(input_node, Node)
+            input_qspec_map[input_node] = get_input_act_qspec(quantization_config)
+
+            weight_node = conv_node.args[1]
+            assert isinstance(weight_node, Node)
+            input_qspec_map[weight_node] = get_weight_qspec(quantization_config)
+
+            bias_node = conv_node.args[2]
+            if isinstance(bias_node, Node):
+                input_qspec_map[bias_node] = get_bias_qspec(quantization_config)
+
+            conv_node.meta["quantization_annotation"] = QuantizationAnnotation(
+                input_qspec_map=input_qspec_map,
+                # TODO<leslie> Remove the annotate of output when oneDNN qconv support fp32 out.
+                output_qspec=get_output_act_qspec(quantization_config),
+                _annotated=True
+            )
```
these seems more or less exact copyies of quantizer we have. we were planning to refactor tis into common utils @jerryzh168

jerryzh168 replied:
yeah, I feel this is somewhat tied to quantization_config, we need to figure out the plan for it

leslie-fang-intel replied:
I think so. One difference is that, we may remove the annotation of `output_qspec` for conv2d in the near future when the backend support it.

kimishpatel commented on torch/ao/quantization/_pt2e/quantizer/x86_inductor_quantizer.py:
```diff
@@ -0,0 +1,205 @@
+
+
+@functools.lru_cache
+def get_default_x86_inductor_quantization_config():
```
So what is really different in this quantizer? Is it qspec?

leslie-fang-intel replied:
Did you mean the difference between `X86InductorQuantizer` and `QNNPackQuantizer`? If so, I think: 1. the supported fusion patterns are different. 2. We will enable some quantization propagation instead of just pattern match for the quantization recipes.

leslie-fang-intel replied:
Sure, in general, after we annotate the fusion patterns of conv/linear, we will go through the graph from start to end and checking if a single quantizable node (for example: relu) is connected to a quantized node of not. If so, we can also annotate this quantizable node as quantized node. We can share more details offline if you are interested.

kimishpatel replied:
> quantized node

Yes I am interested in understanding this a bit better. If you can share more details that would be great. @jerryzh168 whats the best way to do this? If you are allowed to share here, thats even better since we dont have to find a separate communication channel

jerryzh168 replied:
@kimishpatel I posted their IPEX recipe internally in the working group, see the end of the slides

leslie-fang-intel committed:
Update on "[Quant][PT2E]Enable x86 inductor quantizer"


**Summary**

- Enable `X86InductorQuantizer` basics. 
- Recipe to annotate conv2d is added.

**Test Plan**
```
python -u -m pytest -s -v test_quantize_pt2e.py -k TestX86InductorQuantizePT2E
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Enable x86 inductor quantizer"


**Summary**

- Enable `X86InductorQuantizer` basics. 
- Recipe to annotate conv2d is added.

**Test Plan**
```
python -u -m pytest -s -v test_quantize_pt2e.py -k TestX86InductorQuantizePT2E
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Enable x86 inductor quantizer"


**Summary**

- Enable `X86InductorQuantizer` basics. 
- Recipe to annotate conv2d is added.

**Test Plan**
```
python -u -m pytest -s -v test_x86inductor_quantizer.py -k TestX86InductorQuantizePT2E
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant][PT2E]Enable x86 inductor quantizer"


**Summary**

- Enable `X86InductorQuantizer` basics. 
- Recipe to annotate conv2d is added.

**Test Plan**
```
python -u -m pytest -s -v test_x86inductor_quantizer.py -k TestX86InductorQuantizePT2E
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

jerryzh168 commented on test/test_quantization.py:
```diff
@@ -87,6 +87,7 @@
     from quantization.pt2e.test_quantize_pt2e import TestQuantizePT2E  # noqa: F401
     from quantization.pt2e.test_quantize_pt2e import TestQuantizePT2EOps  # noqa: F401
     from quantization.pt2e.test_quantize_pt2e import TestQuantizePT2EModels  # noqa: F401
+    from quantization.pt2e.test_x86inductor_quantizer import TestX86InductorQuantizePT2E  # noqa: F401
```
maybe TestQuantizePT2EX86Inductor?

leslie-fang-intel replied:
Thanks for the suggestion. Change the name to `TestQuantizePT2EX86Inductor`.

jerryzh168 reviewed (approved):
looks good I think, as long as you plan to keep up with our updates of utility functions and quantizer APIs

leslie-fang-intel committed:
Update on "[Quant][PT2E]Enable x86 inductor quantizer"


**Summary**

- Enable `X86InductorQuantizer` basics. 
- Recipe to annotate conv2d is added.

**Test Plan**
```
python -u -m pytest -s -v test_x86inductor_quantizer.py -k TestX86InductorQuantizePT2E
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel commented:
> looks good I think, as long as you plan to keep up with our updates of utility functions and quantizer APIs

Thanks, sure. We will keep up the updates of utility functions and quantizer APIs.

leslie-fang-intel committed:
Update on "[Quant][PT2E]Enable x86 inductor quantizer"


**Summary**

- Enable `X86InductorQuantizer` basics. 
- Recipe to annotate conv2d is added.

**Test Plan**
```
python -u -m pytest -s -v test_x86inductor_quantizer.py -k TestQuantizePT2EX86Inductor
```

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 Xia-Weiwen

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
